bionic/protocols.py

```diff
@@ -599,6 +600,86 @@ def read(self, path):
         return single_element(path.iterdir())
 
 
+# This protocol allows us to hash sets deterministically.
+# It would be preferable if PicklableProtocol could install a special handler for sets,
+# but unfortunately pickle's override mechanisms (dispatch tables and reduce_override)
+# don't work for built-in types like set -- at least not in the C implementation of
+# pickle. This custom protocol will work for set objects, but not for any other objects
+# that happen to contain sets.
+class PicklableSetProtocol(BaseProtocol):
+    """
+    Decorator indicating that an entity's value will be a set or frozenset whose
+    contents are picklable.
+
+    Sets need special handling because their iteration order is non-deterministic, so
+    the result of pickling them is also non-deterministic, which causes problems when
+    the serialized file is hashed. To resolve this, this protocol pickles each element
+    of a set into a separate file, which can then be hashed deterministically.
+
+    Parameters:
+        pickle_protocol_version: int (default: 4)
+            The pickle serialization protocol to use.
+    """
+
+    def __init__(self, pickle_protocol_version=4):
+        super(PicklableSetProtocol, self).__init__()
+        self._pickle_protocol_version = pickle_protocol_version
+
+    def get_fixed_file_extension(self):
+        return "setpkl"
+
+    def validate(self, value):
+        assert isinstance(value, (set, frozenset))
+
+    def write(self, value, path):
+        # We store each element of the set as a separate file, named after the hash of
+        # its contents; this guarantees that the whole directory will get a consistent
+        # hash value regardless of what order the elements are accessed in.
+        path.mkdir()
+
+        items_path = self.items_dir_sub_path(path)
+        items_path.mkdir()
+
+        for item in value:
+            item_file_path = items_path / "_tmp.pkl"
+            with item_file_path.open("wb") as file_:
+                pickle.dump(item, file_, protocol=self._pickle_protocol_version)
+            # After creating the file, we hash its contents and use that as the new
+            # name. However, it's possible for two objects in a set to generate the same
+            # hash (see `test_set_with_duplicate_pickle_output` for an example) so we
+            # need to add a unique suffix as well.
+            item_file_hash = hexdigest_from_path(item_file_path)
+            suffix = 0
+            while True:
+                new_item_file_path = items_path / f"{item_file_hash}_{suffix}.pkl"
+                if not new_item_file_path.exists():
```

Comment:
This can happen when two objects have the same pickle output, as described in the comment above. That's why we have a loop: so we can keep trying new paths until we find one that doesn't exist yet.

I'll try to make the comment clearer.

Reply:
I ended up rewriting the code in a way that's hopefully clearer.